getspice.py

```diff
@@ -22,135 +20,222 @@
+__author__ = 'JorgeDiazdelRio'
+
+
+# =============================================================================
+# Handling of CSPICE downloads from HTTPS
+# ---------------------------------------
+#    Some Python distributions may be linked to an old version of OpenSSL
+#    which will not let you connect to NAIF server due to recent SSL cert
+#    upgrades on the JPL servers.  Moreover, versions older than
+#    OpenSSL 1.0.1g are known to contain the 'the Heartbleed Bug'.  Therefore
+#    this module provides two different implementations for the function
+#    __download_ssl (which implements the HTTPS GET call to the NAIF server
+#    to download the required CSPICE distribution package).
+#
+def _download_ssl(url):
+    """
+    Returns the CSPICE package located at the NAIF server using the input
+    ``url``.  This fuction is implemented using urllib3 and pyOpenSSL in
+    order to resolve the known issue with OpenSSL MacOS Python
+    distributions before Python3.6.x
+
+    Parameters
+    ----------
+    url: str
+       HTTPS URL where the CSPICE package is located within the NAIF
+       server.
+
+    Returns
+    -------
+    io.BytesIO
+       CSPICE package encapsulated within a io.BytesIO stream.
+
+    Note
+    ----
+       This function is a copy of the downloadSpice function existing in
+       the original getspice.py module provided in the SpiceyPy 2.0.1.dev0
+       (by Andrew Annex), with the required modifications to work within
+       this class.
+    """
+    # Send the request to get the CSPICE package.
+    response = urllib.request.urlopen(url, timeout=10)
+
+    # Convert the response.data to io.BytesIO and store it in local memory.
+    return io.BytesIO(response.read())
+
+if ssl.OPENSSL_VERSION < 'OpenSSL 1.0.1g':
+    import pip
+    pip.main(['install', 'pyOpenSSL'])
+    pip.main(['install', 'urllib3[secure]'])
+
+    # Force urllib3 to use pyOpenSSL
+    import urllib3.contrib.pyopenssl
+    urllib3.contrib.pyopenssl.inject_into_urllib3()
+
+    import certifi
+    import urllib3
+
+    def _download_ssl(url):
+        """
+        Returns the CSPICE package located at the NAIF server using the input
+        ``url``.  This fuction is implemented using urllib3 and pyOpenSSL in
+        order to resolve the known issue with OpenSSL MacOS Python
+        distributions before Python3.6.x
+
+        Parameters
+        ----------
+        url: str
+           HTTPS URL where the CSPICE package is located within the NAIF
+           server.
+
+        Returns
+        -------
+        io.BytesIO
+           CSPICE package encapsulated within a io.BytesIO stream.
+        """
+        # Create a PoolManager
+        https = urllib3.PoolManager(cert_reqs='CERT_REQUIRED',
+                                    ca_certs=certifi.where())
+
+        # Send the request to get the CSPICE package.
+        response = https.request('GET', url, timeout=urllib3.Timeout(10))
```

Comment:
I had some code that did download retries that I think is worth keeping around...